LG-10115 - Document IdV rate limiting (#471)
18F/identity-handbook: _articles/identity-proofing-rate-limiting.md

---

> title: "IdV Rate Limiting"

Author: Changed.

description: "An overview of the rate limits which apply to IdV"
layout: article
category: "Team"

> subcategory: People Ops

Author: Changed to AppDev / Architecture. We can discuss further changes and/or moving it.

cSpell: ignore Hubspot
---

> This page lists the rate limits relating to IdV, and provides details

Author: Changed.

about each of them. 

The following IdV related rate limits exist:

- Document Capture (`:idv_doc_auth`)
- Verify Info (`:idv_resolution`)
- Hybrid Handoff (`:idv_send_link`)
- Phone Confirmation (`:proof_address`)
- SSN (`:proof_ssn`)
- GPO (verify by USPS mail) letters
- One-time (SMS) password entry

# Rate Limit Details
## `:idv_doc_auth` - Document capture rate limiter
### Description
This is the rate-limit for the user's attempts to upload their ID
documents (from either their computer or phone). By default, the user
is allowed 5 attempts within 6 hours.
### Config Settings
- `doc_auth_max_attepts` - The maximum number of times the user can
attempt to upload their documents within the specified time window.

- `doc_auth_attempts_window_in_minutes` - The length of time to
  consider when determining whether a user is rate-limited.

### How to trigger
Repeatedly fail doc auth (5 times). This can be done by using a
suitable `.yml` file rather than an image file during image
upload. A suitable file is:

```
failed_alerts:
  - name: Document Classification
    result: Attention```
```

You will need to cancel out of doc auth and then restart it each time;

> the UI won't let you use the same file again if you just retry.

Reviewer: You need two different files, but you don't need to cancel out of doc auth. Suggest linking to [where people can learn more](https://developers.login.gov/testing/#testing-identity-proofing) about yml files to upload. It's worth saying that whether you retry with a different file or cancel and restart, failed attempts still count against the limit within 6 hours.

Author: Point. Got tunnel visioned by how I was doing it. Updating.

Author: Updated

### UI effects

After failing for the final time, the user will be redirected to a
rate limited screen. Any attempt to re-enter IdV will also
redirect there.

![Failed and Rate Limited]({{site.baseurl}}/images/idv-doc-auth-limited.png)

## `:idv_resolution` - Verify info rate limiter
### Description

> This is the rate-limit for the final step in IdV: when the user

Reviewer: It's not the final step. This is the Verify Info rate limiter, that comes before the Phone/Address step.

attempts to submit everything and verify their identity. By default,
the user is allowed 5 attempts in 6 hours.

### Config Settings

- `idv_max_attempts` - The maximum number of times that a user can
attempt to verify their identity documents within the
specified window.

- `idv_attempt_window_in_hours` - The length of time to consider when
 determining whether a user is rate-limited.

### How to trigger
Repeatedly fail IdV. Use a mix of bogus SSNs and device failures, to

> avoid triggering the SSN rate limiter.

Reviewer: What do you mean by device failures? The SSN rate limiter won't be triggered because its limit is double the resolution limit.

### UI effects

After failing for the final time, the user will be redirected to a
rate limited screen. Any attempt to re-enter IdV will also
redirect there.

![Rate Limited]({{site.baseurl}}/images/idv-doc-auth-limited.png)

## `:idv_send_link` - User phone submission of their ID documents (hybrid handoff) rate limiter
### Description
This is the rate limiter for hybrid handoff, where we allow the user
to upload their ID documents from their phone. By default, the user is

> allowed 10 attempts in 5 minutes.

Reviewer: It's 5 attempts in 10 minutes.

### Settings
`idv_send_link_max_attempts` - The maximum number of times that a user can
attempt to upload their ID documents from their phone within the
specified time window.

`idv_send_link_attempt_window_in_minutes` - The length of time to
consider when determining whether a user is rate-limited.

### How to trigger
Enter identity verification, and select hybrid handoff ('Use your

> phone to take photos'). Cancel out of the next screen and return to

Reviewer: ```suggestion
phone to take photos'). Click `Back` on the next screen and return to
```

'How would you like to add your ID?'. Repeat until you become rate
limited.

### UI effects
The user will be presented with a flash error message every time they
attempt to enter hybrid handoff.

[Rate Limited Hybrid Handoff]({{site.baseurl}}/images/hybrid-handoff-limited.png)

## `:proof_address` - Phone verification rate limiter
### Description
This is the rate limiter for phone verification, despite the name. By
default, the user is allowed 5 attempts in 6 hours.

### Config Settings
- `proof_address_max_attempts` - The maximum number of times that a
user can attempt to verify their phone number within the specified
window.

- `proof_address_max_attempt_window_in_minutes` - The length of time
to consider when determining whether a user is rate-limited.

### How to trigger

> Repeatedly fail IdV. Use a mix of bad information (bad SSN, wrong one
> time code, etc.), to avoid triggering the other rate limiters.

Reviewer: I don't think your suggestion would work.
```suggestion
Use phone number 703-555-5555.
```

### UI effects
After failing for the final time, the user will logged out, and their
account will be locked until the login.gov team takes manual action to
re-enable it.

> [Rate limited by address]({{site.baseurl}}/images/idv-address-rate-limited.png)

Reviewer: If you trigger the phone rate limit with the number above, it's the same as the other rate limiters - you can try again after the time period is over.
![Phone rate limit](https://github.com/18F/identity-handbook/assets/2381438/3f7535a6-c06b-401d-9d82-370b81a9c82f)

## `:proof_ssn` - Social Security Number verification rate limiter
### Description
This is the rate limiter for SSN verification. By default, the user is
allowed 10 attempts in 60 minutes.

### Settings
- `proof_ssn_max_attempts` - The maximum number of times that a user
can attempt to verify their social security number within the
specified window.

- `proof_ssn_max_attempt_window_in_minutes` -  The length of time
to consider when determining whether a user is rate-limited.

### How to trigger
Enter IdV. Proceed to the 'Enter your Social Security number'
screen. Repeatedly enter a bad SSN and click Continue until you are

> rate limited.

Reviewer: You have to use two different users (or maybe three) because the resolution rate limit happens first.

### UI effects
After failing for the final time, the user will be redirected to the
IdV rate limited screen. Any attempt to re-enter IdV will also
redirect there.

[Rate Limited]({{site.baseurl}}/images/idv-ssn-rate-limited.png)
## `GpoMail` - GPO (USPS) letters rate limiter

> ### Description

Reviewer: Please use Verify by Mail everywhere instead of GPO.

This is the rate limiter for a user's requests for GPO letters.  This
is a completely independent set of rate limiting code, which
implements two rate limits.

First, a user is not allowed more than a certain number of GPO letter
requests within a time window (this is similar to the other rate
limiters described here).

Second, there is a minimum wait period after requesting a GPO letter
before a user is allowed to request another.

By default, a user is allowed to request 4 GPO letters within 30 days,
and must wait 24 hours after requesting a GPO letter before they are
allowed to request another. (n.b. - in the test environment, users are
restricted to 2 GPO letters per 30 day window; the delay is still 24

> hours.)

Reviewer: What do you mean by test environment? Locally we can override the settings to whatever we need.

### Settings
`max_mail_events` - The maximum number of times that a user may
request a GPO letter within the specified time window.

`max_mail_events_window_in_days` - The length of time to consider when
determining whether the user has requested too many GPO letters
recently.

`minimum_wait_before_another_usps_letter_in_hours` - The minimum
amount of time that a user must wait, after requesting a GPO letter,
before requesting another letter.

### How to trigger
Enter IdV and select GPO letter address verification. Request a GPO
letter; you are now rate-limited.
### UI effects

> On the 'Welcome Back' screen, the user is not presented with the

Reviewer: ```suggestion
On the screen to enter their verification code, the user is not presented with the
```

option to request another letter.

[Rate limited]({{site.baseurl}}/images/gpo_letter_request_rate_limited.png)

## `User` OTP verification rate limiter
When the user is entering a one-time (SMS) password, rate limiting is
handled by a custom set of code. The actual rate limits are stored on
`User`. This is strictly for attempts to enter the one-time password.

If the user fails all of their attempts, they are logged out and must
wait a while before we allow them to log in again.

By default, the user is allowed 10 attempts. If they are locked out,
they must wait 10 minutes before we allow them to log back in.

(n.b. - This code is _also_ used for rate-limiting OTP entry during

> login.)

Reviewer: I thought you were talking about MFA during login. What do you mean by one-time (SMS) password above? Are you talking about the Phone step?

---
title: "Identity Verification Rate Limiting"
description: "An overview of the rate limits which apply to Identity Verification"
layout: article
category: AppDev
subcategory: Architecture
cSpell: ignore Hubspot
---

This page lists the rate limits relating to Identity Verification, and provides details
about each of them. 

The following Identity Verification related rate limits exist:

> - Document Capture (`:idv_doc_auth`)

Reviewer: It would be great to make these link internally to each section.

- Verify Info (`:idv_resolution`)
- Hybrid Handoff (`:idv_send_link`)
- Phone Confirmation (`:proof_address`)
- SSN (`:proof_ssn`)
- Verify by Mail (verify by USPS mail) letters

> - One-time (SMS) password entry

Reviewer: If you're referring to MFA for logging in here, this isn't IdV-related.

# Rate Limit Details
## `:idv_doc_auth` - Document capture rate limiter
### Description
This is the rate-limit for the user's attempts to upload their ID
documents from either their computer or phone. By default, the user
is allowed 5 attempts within 6 hours.

Retrying with a different file after a failure and cancelling out of
Identity Verification completely and then trying again both count
against this rate limiter. The user gets 5 attempts to upload

> satisfactory images in 6 hours total, through any path.

Reviewer: Both successful and failed attempts are now counted for doc_auth, resolution, ssn, and proof_address rate limiters. So some of the detailed info we added about how to fail these various checks isn't needed anymore (although useful to have), since successful attempts are also counted.

### Config Settings

> - `doc_auth_max_attepts` - The maximum number of times the user can

Reviewer: ```suggestion
- `doc_auth_max_attempts` - The maximum number of times the user can
```

attempt to upload their documents within the specified time window.

- `doc_auth_attempts_window_in_minutes` - The length of time to
  consider when determining whether a user is rate-limited.

### How to trigger
Repeatedly fail doc auth (5 times). This can be done by using a
suitable `.yml` file rather than an image file during image
upload.

Use the same file for front and back images, but you must alternate

> between two files on succesive attempts. The UI will not let you try

Reviewer: ```suggestion
between two files on successive attempts. The UI will not let you try
```

phone number 703-555-5555. Identity verification will fail. Retry,
using the same phone number, until you are rate-limited. Ignore the
message about using a different phone number.

### UI effects

The user will be redirected to a screen informing them that they are
rate-limited. Any further attempt to proof their address before the
rate limit expires will also be directed to this screen.

![Proof Address Rate Limited]({{site.baseurl}}/images/idv-proof-address-rate-limited.png)

## One time password entry rate limiter
### Description
During the phone confirmation step, the user must enter a one-time
code sent to their phone. This is the rate limiter for requesting
one-time codes.

When the user is entering a one-time code, rate limiting is
handled by code which is common to both OTP entry during
identity verification _and_ during user login.

The actual rate limit count is stored on the `User` class. Most of the
code is in the `UserOtpMethods` concern, with a small bit of it still
in `User`.

By default, the user is allowed to request 10 one-time passwords
within 10 minutes.

If the user requests more than 10, they are blocked from further

> access, and must wait 10 minutes before we allow them to continue.

Reviewer: ```suggestion
access, and must wait 10 minutes before being allowed to continue.
```
